File: leetcode_python/maximum_subarray.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def maxSubArray(self, A):
        max_ending_here = max_so_far = A[0]
        for x in A[1:]:
            max_ending_here = max(max_ending_here + x, x)
            max_so_far = max(max_so_far, max_ending_here)
        return max_so_far

File: leetcode_cpp/maximum_subarray.cpp
class Solution {
public:
    int maxSubArray(int A[], int n) {
        int max_ending_here = A[0], max_so_far = A[0];
        for (int i = 1; i < n; ++i) {
            max_ending_here = max(max_ending_here + A[i], A[i]);
            max_so_far = max(max_so_far, max_ending_here);
        }
        return max_so_far;
    }
};
